reframe/core/pipeline.py

```diff
@@ -930,35 +947,66 @@ def compile(self, **compile_opts):
+                self.build_system = 'SingleSource'
+
+            self.build_system.srcfile = self.sourcepath
+            self.build_system.executable = self.executable
+
+        if compile_opts:
+            user_deprecation_warning(
+                'passing options to the compile() method is deprecated; '
+                'please use a build system.')
+
+            # Remove source and executable from compile_opts
+            compile_opts.pop('source', None)
+            compile_opts.pop('executable', None)
+            try:
+                self.build_system.makefile = compile_opts['makefile']
+                self.build_system.options  = compile_opts['options']
+            except KeyError:
+                pass
+
+        # Prepare build job
+        build_commands = [
+            *self.prebuild_cmd,
+            *self.build_system.emit_build_commands(self._current_environ),
+            *self.postbuild_cmd
+        ]
+        self._build_job = getscheduler('local')(
+            name='rfm_%s_build' % self.name,
+            launcher=getlauncher('local')(),
+            command='\n'.join(build_commands),
+            environs=[self._current_partition.local_env,
+                      self._current_environ],
+            workdir=self._stagedir)
 
-        # Change working dir to stagedir although absolute paths are used
-        # everywhere in the compilation process. This is done to ensure that
-        # any other files (besides the executable) generated during the the
-        # compilation will remain in the stage directory
         with os_ext.change_dir(self._stagedir):
-            self.prebuild()
-            if os.path.isdir(staged_sourcepath):
-                includedir = staged_sourcepath
-            else:
-                includedir = os.path.dirname(staged_sourcepath)
-
-            self._current_environ.include_search_path.append(includedir)
-            self._compile_proc = self._current_environ.compile(
-                sourcepath=staged_sourcepath,
-                executable=os.path.join(self._stagedir, self.executable),
-                **compile_opts)
-            self.logger.debug('compilation stdout:\n%s' %
-                              self._compile_proc.stdout)
-            self.logger.debug('compilation stderr:\n%s' %
-                              self._compile_proc.stderr)
-            self.postbuild()
+            try:
+                self._build_job.prepare(BashScriptBuilder())
+            except OSError as e:
+                raise PipelineError('failed to prepare build job') from e
 
+            self._build_job.submit()
+
+    def compile_wait(self):
+        """Wait for compilation phase to finish.
+
+        .. versionadded:: 2.13
+        """
+        self._build_job.wait()
         self.logger.debug('compilation finished')
 
+        # FIXME: this check is not reliable for certain scheduler backends
+        if self._build_job.exitcode != 0:
+            raise PipelineError('compilation failed')
```

Comment:
I should perhaps raise a more specific error here, e.g., `BuildError`. What do you think?